### PR TITLE
#816 compile heron with bazel 0.2.3

### PR DESCRIPTION
**Pull request by caofangkun**
Since Bazel Release 0.1.5 (2016-02-05)	

"Repository rules must use names that are valid workspace names.", see: https://github.com/bazelbuild/bazel/issues/874

```
➜  heron git:(master) ✗ bazel version 
Build label: 0.2.3
Build target: bazel-out/local-fastbuild/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Tue May 17 14:21:13 2016 (1463494873)
Build timestamp: 1463494873
Build timestamp as int: 1463494873
```

### CODE REVIEW

File: 3rdparty/java/BUILD [hunk 1]
```diff
@@ -6,90 +6,90 @@ java_library(
     name = "aws-java-sdk",
```
Comment (kramasamy) [inline 3rdparty/java/BUILD]: Thanks for the fix. In order for the travis ci to compile correctly, you need to download bazel 0.2.3 in the .travis.yml 
https://github.com/twitter/heron/blob/master/.travis.yml

Can you make the changes and update the PR?

Comment (caofangkun) [inline 3rdparty/java/BUILD]: Hi @kramasamy I have fixed this bug. 

The Travis CI still not pass yet, I will  get into the root of the matter to find out why.

**Comment by kramasamy** [comment]
@caofangkun - When I looked the log of travis ci, it seems to failing in pants. I was wondering if you could enable some logs so that we can identify why the error message is being issued. 

Based on @billonahill comments - I recommend they add some debugging to see what file isn't being copied, or to set should_log to return True in tracer.py

**Comment by taishi8117** [comment]
@caofangkun We've figured out that Bazel made some internal changes with handling Python scripts, which seem to have caused the travis failure. The following diff should fix this issue.

```
diff --git a/3rdparty/pex/_pex.py b/3rdparty/pex/_pex.py
index 14e1ffe..a0c4f3b 100644
--- a/3rdparty/pex/_pex.py
+++ b/3rdparty/pex/_pex.py
@@ -20,7 +20,7 @@ if not zipfile.is_zipfile(sys.argv[0]):
     sys.modules.pop('twitter.common.python', None)

     root = os.path.join(
-        os.sep.join(__file__.split(os.sep)[:-6]), '3rdparty/pex/_pex.runfiles/3rdparty')
+        os.sep.join(__file__.split(os.sep)[:-6]), 'pex/_pex.runfiles/__main__/3rdparty')
     sys.path.insert(0, os.path.join(root, 'pex'))
     sys.path.insert(0, os.path.join(root, 'setuptools'))
     setuptools_py =  os.path.join(root, 'setuptools')
```

Additionally, the following modifications may be necessary.

```
diff --git a/scripts/packages/BUILD b/scripts/packages/BUILD
index a990e75..4779d77 100644
--- a/scripts/packages/BUILD
+++ b/scripts/packages/BUILD
@@ -1,7 +1,6 @@
 package(default_visibility = ["//visibility:public"])

-load("/tools/build_defs/pkg/pkg", "pkg_tar")
-load("/tools/build_defs/pkg/pkg", "pkg_deb")
+load("@bazel_tools//tools/build_defs/pkg:pkg.bzl", "pkg_tar", "pkg_deb")
```

```
diff --git a/tools/java/BUILD b/tools/java/BUILD
index 3a099b8..74a39b4 100644
--- a/tools/java/BUILD
+++ b/tools/java/BUILD
@@ -9,7 +9,7 @@ java_toolchain(
     target_version = "7",
     bootclasspath = ["@bazel_tools//tools/jdk:bootclasspath"],
     extclasspath = ["@bazel_tools//tools/jdk:extdir"],
-    javac = ["@bazel_tools//tools/jdk:javac"],
+    javac = ["@bazel_tools//third_party/java/jdk/langtools:javac_jar"],
     javabuilder = ["@bazel_tools//tools/jdk:JavaBuilder_deploy.jar"],
     singlejar = ["@bazel_tools//tools/jdk:SingleJar_deploy.jar"],
     genclass = ["@bazel_tools//tools/jdk:GenClass_deploy.jar"],
```

**Comment by kramasamy** [comment]
I am also getting this warnings - 

```
bazel build --config=ubuntu heron/...
WARNING: /Users/kramasamy/workspace/heron24/3rdparty/zookeeper/BUILD:82:16: in includes attribute of cc_library rule //3rdparty/zookeeper:zookeeper-cxx: 'include' resolves to '3rdparty/zookeeper/include' not in 'third_party'. This will be an error in the future.
WARNING: /Users/kramasamy/workspace/heron24/3rdparty/libevent/BUILD:100:16: in includes attribute of cc_library rule //3rdparty/libevent:libevent-cxx: 'include' resolves to '3rdparty/libevent/include' not in 'third_party'. This will be an error in the future.
WARNING: /Users/kramasamy/workspace/heron24/3rdparty/gflags/BUILD:49:16: in includes attribute of cc_library rule //3rdparty/gflags:gflags-cxx: 'include' resolves to '3rdparty/gflags/include' not in 'third_party'. This will be an error in the future.
WARNING: /Users/kramasamy/workspace/heron24/3rdparty/gtest/BUILD:94:16: in includes attribute of cc_library rule //3rdparty/gtest:gtest-cxx: 'include' resolves to '3rdparty/gtest/include' not in 'third_party'. This will be an error in the future.
WARNING: /Users/kramasamy/workspace/heron24/3rdparty/gperftools/BUILD:71:16: in includes attribute of cc_library rule //3rdparty/gperftools:tcmalloc-cxx: 'include' resolves to '3rdparty/gperftools/include' not in 'third_party'. This will be an error in the future.
WARNING: /Users/kramasamy/workspace/heron24/3rdparty/jansson/BUILD:46:16: in includes attribute of cc_library rule //3rdparty/jansson:jansson-cxx: 'include' resolves to '3rdparty/jansson/include' not in 'third_party'. This will be an error in the future.
WARNING: /Users/kramasamy/workspace/heron24/3rdparty/gmock/BUILD:83:16: in includes attribute of cc_library rule //3rdparty/gmock:gmock-cxx: 'include' resolves to '3rdparty/gmock/include' not in 'third_party'. This will be an error in the future.
WARNING: /Users/kramasamy/workspace/heron24/3rdparty/glog/BUILD:54:16: in includes attribute of cc_library rule //3rdparty/glog:glog-cxx: 'include' resolves to '3rdparty/glog/include' not in 'third_party'. This will be an error in the future.
WARNING: /Users/kramasamy/workspace/heron24/3rdparty/protobuf/BUILD:141:16: in includes attribute of cc_library rule //3rdparty/protobuf:protobuf-cxx: 'include' resolves to '3rdparty/protobuf/include' not in 'third_party'. This will be an error in the future.
WARNING: /Users/kramasamy/workspace/heron24/3rdparty/yaml-cpp/BUILD:127:16: in includes attribute of cc_library rule //3rdparty/yaml-cpp:yaml-cxx: 'include' resolves to '3rdparty/yaml-cpp/include' not in 'third_party'. This will be an error in the future.
```

**Comment by kramasamy** [comment]
When I looked at the travis-ci logs, it seems to be failing with java tests.

**Comment by caofangkun** [comment]
Since Bazel 0.2.3, Bazel  warns if a cc rule's includes attribute points out of third_party
https://github.com/bazelbuild/bazel/commit/f0a18846461d4b7ecc76c36a659ff1e37f93d7dc

**Comment by taishi8117** [comment]
Bazel team is working on this issue -- they might remove these warnings as well.
bazelbuild/bazel#1286

**Comment by taishi8117** [comment]
For unit tests failure, it was because of Bazel's change in java_test behavior since 0.2.1.
bazelbuild/bazel#1017

**Comment by taishi8117** [comment]
@caofangkun I've fixed unit test failures for bazel 0.2.3 and pushed to my branch taishi8117/heron@7bffc5a08e44dfc6bc62425b5f75d5f9dd9aec02.
It would be great if you could merge it.

**Comment by taishi8117** [comment]
@caofangkun Since Bazel's internal test structure changed, the following changes are necessary, as well as pull from #938 (already merged to master)

```
diff --git a/heron/common/tests/java/com/twitter/heron/common/config/Constants.java b/heron/common/tests/java/com/twitter/heron/common/config/Constants.java
index bd13b68..db1ed1e 100644
--- a/heron/common/tests/java/com/twitter/heron/common/config/Constants.java
+++ b/heron/common/tests/java/com/twitter/heron/common/config/Constants.java
@@ -24,7 +24,7 @@ public final class Constants {
   public static final String LAUNCHER_CLASS_KEY = "heron.launcher.class";
 
   public static final String TEST_DATA_PATH =
-      "/heron/common/tests/java/com/twitter/heron/common/config/testdata";
+      "/__main__/heron/common/tests/java/com/twitter/heron/common/config/testdata";
 
   private Constants() {
   }
diff --git a/heron/instance/tests/java/com/twitter/heron/resource/Constants.java b/heron/instance/tests/java/com/twitter/heron/resource/Constants.java
index cd8b8d1..dadef82 100644
--- a/heron/instance/tests/java/com/twitter/heron/resource/Constants.java
+++ b/heron/instance/tests/java/com/twitter/heron/resource/Constants.java
@@ -41,7 +41,7 @@ public final class Constants {
   // For bazel, we use the env var to get the path of heron internals config file
   public static final String BUILD_TEST_SRCDIR = "TEST_SRCDIR";
   public static final String BUILD_TEST_HERON_INTERNALS_CONFIG_PATH =
-      "/heron/config/src/yaml/conf/test/test_heron_internals.yaml";
+      "/__main__/heron/config/src/yaml/conf/test/test_heron_internals.yaml";
 
   private Constants() {
   }
diff --git a/heron/spi/tests/java/com/twitter/heron/spi/common/TestConstants.java b/heron/spi/tests/java/com/twitter/heron/spi/common/TestConstants.java
index f968363..cc2cd59 100644
--- a/heron/spi/tests/java/com/twitter/heron/spi/common/TestConstants.java
+++ b/heron/spi/tests/java/com/twitter/heron/spi/common/TestConstants.java
@@ -16,7 +16,7 @@ package com.twitter.heron.spi.common;
 
 final class TestConstants {
   public static final String TEST_DATA_PATH =
-      "/heron/spi/tests/java/com/twitter/heron/spi/common/testdata";
+      "/__main__/heron/spi/tests/java/com/twitter/heron/spi/common/testdata";
 
   private TestConstants() {
   }
diff --git a/heron/uploaders/tests/java/com/twitter/heron/uploader/localfs/LocalFileSystemConstantsTest.java b/heron/uploaders/tests/java/com/twitter/heron/uploader/localfs/LocalFileSystemConstantsTest.java
index d206f33..deb96cd 100644
--- a/heron/uploaders/tests/java/com/twitter/heron/uploader/localfs/LocalFileSystemConstantsTest.java
+++ b/heron/uploaders/tests/java/com/twitter/heron/uploader/localfs/LocalFileSystemConstantsTest.java
@@ -16,7 +16,7 @@ package com.twitter.heron.uploader.localfs;
 
 final class LocalFileSystemConstantsTest {
   public static final String TEST_DATA_PATH =
-      "/heron/uploaders/tests/java/com/twitter/heron/uploader/localfs/testdata";
+      "/__main__/heron/uploaders/tests/java/com/twitter/heron/uploader/localfs/testdata";
 
   private LocalFileSystemConstantsTest() {
   }
```

**Comment by caofangkun** [comment]
Hi @kramasamy @taishi8117 
Thank you very much for your kind help. CI passed ... at last!

**Comment by kramasamy** [comment]
Awesome! The only issue remaining is the warnings which is not fixed even in bazel 0.3.0

**Comment by caofangkun** [comment]
Hi @kramasamy 

According to code of https://github.com/bazelbuild/bazel/commit/f0a18846461d4b7ecc76c36a659ff1e37f93d7dc#diff-b1e05be607e9de9e40f306162531dc81R431

I did the following changes, and the warnings gone.

```
twitter-heron git:(heron-816-1) ✗ mv 3rdparty third_party
twitter-heron git:(heron-816-1) ✗ sed -i "s/3rdparty/third_party/g" `grep 3rdparty -rl ./`
```

**Comment by billonahill** [comment]
Please update bazel version for the website here too:
https://github.com/twitter/heron/blob/master/website/config.yaml#L24

And in the md files we should reference it via:
```
{{% bazelVersion %}}
```

File: website/content/docs/developers/compiling/docker.md [hunk 2]
```diff
@@ -162,11 +162,11 @@ RUN \
 ENV JAVA_HOME /usr/lib/jvm/java-8-oracle
 ```
 
-#### Step 5 - An installation script for [Bazel](http://bazel.io/) version 0.1.2 or above.
+#### Step 5 - An installation script for [Bazel](http://bazel.io/) version 0.2.3 or above.
```
Comment (billonahill) [inline website/content/docs/developers/compiling/docker.md]: Please update to use `{{% bazelVersion %}}` here and below.

File: website/content/docs/developers/compiling/linux.md [hunk 3]
```diff
@@ -176,12 +176,12 @@ export JRE_HOME=/opt/jdk1.8.0_91/jre
 export PATH=$PATH:/opt/jdk1.8.0_91/bin:/opt/jdk1.8.0_91/jre/bin
 ```
 
-#### Step 5 - Install Bazel 0.1.2
+#### Step 5 - Install Bazel 0.2.3
```
Comment (billonahill) [inline website/content/docs/developers/compiling/linux.md]: Please update to use `{{% bazelVersion %}}` here, above and below.

File: website/content/docs/developers/compiling/mac.md [hunk 4]
```diff
@@ -28,9 +28,9 @@ $ echo $CC $CXX
 #### Step 4 - Install Bazel
 
 ```bash
-curl -O -L https://github.com/bazelbuild/bazel/releases/download/0.1.2/bazel-0.1.2-installer-darwin-x86_64.sh
-chmod +x bazel-0.1.2-installer-darwin-x86_64.sh
-./bazel-0.1.2-installer-darwin-x86_64.sh --user
+curl -O -L https://github.com/bazelbuild/bazel/releases/download/0.2.3/bazel-0.2.3-installer-darwin-x86_64.sh
```
Comment (billonahill) [inline website/content/docs/developers/compiling/mac.md]: Please update to use `{{% bazelVersion %}}`.

**Comment by kramasamy** [comment]
@caofangkun - this PR is really coming out very well. I see some additional warnings - 

```
INFO: From Executing extra_action //tools/cpp:checkstyle_cpp on //heron/common/tests/cpp/threads:spcountdownlatch_unittest:
Ignoring external/local_config_cc/cc_wrapper.sh; not a valid file name (cc, h, cpp, cu, cuh)
INFO: From Executing extra_action //tools/cpp:checkstyle_cpp on //heron/proto:proto-cxx:
Ignoring external/local_config_cc/cc_wrapper.sh; not a valid file name (cc, h, cpp, cu, cuh)
INFO: From Executing extra_action //tools/cpp:checkstyle_cpp on //heron/common/src/cpp/network:network-cxx:
Ignoring external/local_config_cc/cc_wrapper.sh; not a valid file name (cc, h, cpp, cu, cuh)
INFO: From Executing extra_action //tools/cpp:checkstyle_cpp on //heron/common/src/cpp/network:network-cxx:
Ignoring external/local_config_cc/cc_wrapper.sh; not a valid file name (cc, h, cpp, cu, cuh)
INFO: From Executing extra_action //tools/cpp:checkstyle_cpp on //heron/common/src/cpp/network:network-cxx:
Ignoring external/local_config_cc/cc_wrapper.sh; not a valid file name (cc, h, cpp, cu, cuh)
INFO: From Executing extra_action //tools/cpp:checkstyle_cpp on //heron/common/src/cpp/zookeeper:zookeeper-cxx:
Ignoring external/local_config_cc/cc_wrapper.sh; not a valid file name (cc, h, cpp, cu, cuh)
INFO: From Executing extra_action //tools/cpp:checkstyle_cpp on //heron/common/tests/cpp/network:switch_unittest:
Ignoring external/local_config_cc/cc_wrapper.sh; not a valid file name (cc, h, cpp, cu, cuh)
INFO: From Executing extra_action //tools/cpp:checkstyle_cpp on //heron/proto:proto_execution_state_cc:
Ignoring external/local_config_cc/cc_wrapper.sh; not a valid file name (cc, h, cpp, cu, cuh)
INFO: From Executing extra_action //tools/cpp:checkstyle_cpp on //heron/common/tests/cpp/network:order_unittest:
Ignoring external/local_config_cc/cc_wrapper.sh; not a valid file name (cc, h, cpp, cu, cuh)
INFO: From Executing extra_action //tools/cpp:checkstyle_cpp on //heron/common/tests/cpp/network:switch_unittest:
Ignoring external/local_config_cc/cc_wrapper.sh; not a valid file name (cc, h, cpp, cu, cuh)
INFO: From Executing extra_action //tools/cpp:checkstyle_cpp on //heron/common/tests/cpp/network:switch_unittest:
Ignoring external/local_config_cc/cc_wrapper.sh; not a valid file name (cc, h, cpp, cu, cuh)
INFO: From Executing extra_action //tools/cpp:checkstyle_cpp on //heron/common/tests/cpp/network:packet_unittest:
Ignoring external/local_config_cc/cc_wrapper.sh; not a valid file name (cc, h, cpp, cu, cuh)
INFO: From Executing extra_action //tools/cpp:checkstyle_cpp on //heron/common/tests/cpp/network:order_unittest:
Ignoring external/local_config_cc/cc_wrapper.sh; not a valid file name (cc, h, cpp, cu, cuh)
INFO: From Executing extra_action //tools/cpp:checkstyle_cpp on //heron/common/tests/cpp/network:order_unittest:
Ignoring external/local_config_cc/cc_wrapper.sh; not a valid file name (cc, h, cpp, cu, cuh)
INFO: From Executing extra_action //tools/cpp:checkstyle_cpp on //heron/stmgr/src/cpp:grouping-cxx:
Ignoring external/local_config_cc/cc_wrapper.sh; not a valid file name (cc, h, cpp, cu, cuh)
```

**Comment by caofangkun** [comment]
Hi @kramasamy  
I checked the  travis-ci logs but not find  any ```not a valid file name (cc, h, cpp, cu, cuh)``` warnings.

**Comment by kramasamy** [comment]
@caofangkun - this occurs in my Mac laptop.

File: website/content/docs/developers/compiling/mac.md (hunk 4 above)
Comment (kramasamy) [inline website/content/docs/developers/compiling/mac.md]: @caofangkun - can you include @billonahill feedback on the docs?

Comment (caofangkun) [inline website/content/docs/developers/compiling/mac.md]: @kramasamy OK, I will do it .

**Comment by caofangkun** [comment]
Hi @billonahill I have include your feedback in d74c1dc , Could you please have a review?

**Comment by kramasamy** [comment]
@caofangkun - travis-ci seems to have failed. Can you please check?

**Comment by taishi8117** [comment]
@kramasamy @caofangkun For the `not a valid file name (cc, h, cpp, cu, cuh)` warnings, the following diff should resolve it.

```
diff --git a/tools/java/src/com/twitter/bazel/checkstyle/CppCheckstyle.java b/tools/java/src/com/twitter/bazel/checkstyle/CppCheckstyle.java
index dd5d758..d1b6f87 100644
--- a/tools/java/src/com/twitter/bazel/checkstyle/CppCheckstyle.java
+++ b/tools/java/src/com/twitter/bazel/checkstyle/CppCheckstyle.java
@@ -134,6 +134,7 @@ public final class CppCheckstyle {
                     Predicates.not(Predicates.containsPattern("third_party/")),
                     Predicates.not(Predicates.containsPattern("config/heron-config.h")),
                     Predicates.not(Predicates.containsPattern(".*pb.h$")),
+                    Predicates.not(Predicates.containsPattern(".*cc_wrapper.sh$")),
                     Predicates.not(Predicates.containsPattern(".*pb.cc$"))
             )
     );
```

I'll be looking into the issues with Travis failure as well.

**Comment by taishi8117** [comment]
Also, just realized that there existed the following error even from  [cc5da14](https://github.com/twitter/heron/pull/869/commits/cc5da142af4043af4aa85988a927de3b04375f64), although Travis succeeded for some reasons.

```
ERROR: /home/travis/build/twitter/heron/heron/proto/BUILD:109:1: error executing shell command: 'set -e
rm -rf bazel-out/local-fastbuild/genfiles/heron/proto/proto_stmgr_java_src.srcjar.srcs
mkdir bazel-out/local-fastbuild/genfiles/heron/proto/proto_stmgr_java_src.srcjar.srcs
bazel-out/local-f...' failed: namespace-sandbox failed: error executing command
```

File: website/content/docs/developers/compiling/mac.md [hunk 5]
```diff
@@ -25,12 +25,12 @@ $ export CXX=/usr/bin/clang++
...
 ```
 
-#### Step 4 - Install Bazel
+#### Step 4 - Install Bazel {{% bazelVersion %}}
 
 ```bash
-curl -O -L https://github.com/bazelbuild/bazel/releases/download/0.1.2/bazel-0.1.2-installer-darwin-x86_64.sh
-chmod +x bazel-0.1.2-installer-darwin-x86_64.sh
-./bazel-0.1.2-installer-darwin-x86_64.sh --user
+wget -O /tmp/bazel.sh https://github.com/bazelbuild/bazel/releases/download/{{% bazelVersion %}}/bazel-{{% bazelVersion %}}-installer-linux-x86_64.sh
```
Comment (taishi8117) [inline website/content/docs/developers/compiling/mac.md]: s/linux/darwin/

**Comment by caofangkun** [comment]
Hi @taishi8117 , Thank you so much for your help, I have fixed it as you suggested.

**Comment by kramasamy** [comment]
@caofangkun - can you update your branch since yours seems to have conflicts?

File: website/content/docs/developers/compiling/mac.md (hunk 5 above)
Comment (caofangkun) [inline website/content/docs/developers/compiling/mac.md]: fixed.

**Comment by taishi8117** [comment]
@caofangkun Thanks for fixing conflicts. The Javadoc error can be resolved by merging #968 --- can you try that?

**Comment by caofangkun** [comment]
Thank you @taishi8117 After merge the CI passed now.

**Comment by kramasamy** [comment]
Looks good to me! 👍

**Comment by taishi8117** [comment]
Looks great!

**Comment by nlu90** [comment]
@kramasamy What are the changes need to be made for  our Twitter internal building/testing environment?

**Pull request merged**